src/shared/firewall-util-nft.c

```diff
@@ -0,0 +1,853 @@
+        if (r < 0)
+                return r;
+
+        nfnl = sd_netlink_ref(nfnlp);
+        assert_return(nfnlp, -EINVAL);
+
+        r = sd_nfnl_message_batch_begin(nfnl, &ruleset[msgcnt]);
+        if (r < 0)
+                goto out_unref;
+
+        assert_return(msgcnt < NFT_INIT_MSGS, -EINVAL);
+        r= __fw_nftables_init(ruleset, msgcnt);
+        if (r < 0)
+                goto out_unref;
+
+        assert_return(r > (int)msgcnt, -EINVAL);
+
+        msgcnt = r;
+
+        assert_return(msgcnt < NFT_INIT_MSGS, -EINVAL);
+        r = sd_nfnl_message_batch_end(nfnl, &ruleset[++msgcnt]);
+        if (r < 0)
+                goto out_unref;
+
+        r = nfnl_netlink_sendv(nfnl, ruleset, ++msgcnt);
+        if (r < 0)
+                return r == -EEXIST ? 0 : r;
+
+        return 0;
+out_unref:
+        sd_netlink_unref(nfnl);
+        while (msgcnt)
+                sd_netlink_message_unref(ruleset[--msgcnt]);
+        return r;
+}
+
+static int nft_message_add_setelem_iprange(sd_netlink_message *m,
+                                           const union in_addr_union *source,
+                                           unsigned int prefixlen)
+{
+        union in_addr_union source_end = *source;
+        unsigned int nplen = 32 - prefixlen;
+        uint32_t v = (1 << nplen);
+        int r;
+
+        r = sd_nfnl_nft_message_add_setelem(m, 0, &source->in, sizeof(uint32_t), NULL, 0);
+        if (r < 0)
+                return r;
+        r = sd_nfnl_nft_message_add_setelem_end(m);
+        if (r < 0)
+                return r;
+
+        v = (1 << nplen);
+
+        source_end.in.s_addr += htonl(v);
+
+        r = sd_nfnl_nft_message_add_setelem(m, 1, &source_end.in, sizeof(uint32_t), NULL, 0);
+        if (r < 0)
+                return r;
+
+        r = sd_netlink_message_append_u32(m, NFTA_SET_ELEM_FLAGS, htobe32(NFT_SET_ELEM_INTERVAL_END));
+        if (r < 0)
+                return r;
+
+        r = sd_nfnl_nft_message_add_setelem_end(m);
+        if (r < 0)
+                return r;
+
+        return r;
+}
+
+int fw_nftables_add_masquerade(
+                bool add,
+                int af,
+                const union in_addr_union *source,
+                unsigned source_prefixlen) {
+        sd_netlink_message *transaction[3];
+        int r;
+
+        if (!source || source_prefixlen == 0)
+               return -EOPNOTSUPP;
+
+        memset(transaction, 0, sizeof(transaction));
+
+        r = sd_nfnl_message_batch_begin(nfnl, &transaction[0]);
+        if (r < 0)
+               return r;
+
+        if (add) {
+                 r = sd_nfnl_nft_message_new_setelems_begin(nfnl, &transaction[1], nft_systemd_table_name, "masq_saddr");
+                 if (r < 0)
+                        return r;
+        } else {
+                 r = sd_nfnl_nft_message_del_setelems_begin(nfnl, &transaction[1], nft_systemd_table_name, "masq_saddr");
+                 if (r < 0)
+                        return r;
```

Comment:
do the `if (r < 0) return r` check outside of the `if (add)` branches? it's going to be a bit shorter then!